File: #115_Distinct_Subsequence/Solution.java
public class Solution {
    public int numDistinct(String s, String t) {
        // System.out.println(s + ", " + t);
        int sum = 0;
        int slen = s.length();
        int tlen = t.length();
        if (slen == 0 && tlen == 0) return 1;
        if (slen == 0) return 0;
        if (tlen == 0) return 1;
        if (s.equals(t)) return 1;
        if (tlen > slen) return 0;
        for (int i = 0; i < slen - tlen + 1; i++) {
            System.out.println("i: " + i);
            if (s.charAt(i) == t.charAt(0)) sum += numDistinct(s.substring(i + 1), t.substring(1));
        }
        return sum;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        String s = "ccc";
        String t = "c";
        System.out.println(sol.numDistinct(s, t));
    }
}

File: #115_Distinct_Subsequence/Solution.py
class Solution:
	# @param {string} s
	# @param {string} t
	# @return {integer}
	def numDistinct(self, s, t):
		if not t:
			return 1
		if not s:
			return 0
		if len(s) < len(t):
			return 0
		dp = [[0 for j in range(len(t) + 1)] for i in range(len(s) + 1)]
		for i in range(len(s) + 1):
			dp[i][0] = 1  # initialize, empty t matches all s
		for i in range(1, len(s) + 1):
			for j in range(1, len(t) + 1):
				if j > i: # t is longer than s
					continue
				if t[j - 1] == s[i - 1]:
					dp[i][j] = dp[i - 1][j - 1] # use the char at i
				dp[i][j] += dp[i - 1][j] # delete char at i
		return dp[len(s)][len(t)]

sol = Solution()
print sol.numDistinct("rabbbit", "rabbit")